### PR TITLE
[music] When adding a new music directory, ask the user about the con…

### PR DESCRIPTION
**Pull request by razzeee**
…tent and if he wants to scan it

https://github.com/xbmc/xbmc/pull/8048#issuecomment-140036382 is now obsolete.

This will show the GUIDialogContentSettings (scraper selection), when an user is adding a new directory as music source.
Only tested on windows so far, will need tests on all platforms, as I changed a bunch of path handling stuff.

Kind of technical details:
- This converts VFS paths in the content table to absolute paths
- Disables the context menu entry for genres and folders that might contain albums/artists (just set these when you setup the folders)
- One central place for the GUIDialogContentSettings call handled like in the videodb
- Users can now select at will if they want to use an artist or album scraper
- After the scraper is set, it will ask if you want to scan

### CODE REVIEW

File: xbmc/music/MusicDatabase.cpp [hunk 1]
```diff
@@ -3946,11 +3946,38 @@ void CMusicDatabase::UpdateTables(int version)
       m_pDS->exec("DROP TABLE album");
       m_pDS->exec("ALTER TABLE album_new RENAME TO album");
    }
+  if (version < 55)
+  {
+      // Convert the content table from using musicdb:// paths to actual paths
+      if (NULL == m_pDB.get()) return;
+      if (NULL == m_pDS.get()) return;
+      //Genres are no longer supported here
+      m_pDS->exec(PrepareSQL("SELECT strPath FROM content WHERE strPath LIKE 'musicdb://genres/%'"));
```
Comment (anaconda) [inline xbmc/music/MusicDatabase.cpp]: Is this supposed to be `DELETE FROM`?

Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Yes, will fix.

**Comment by Razzeee** [comment]
Also needs testing on mysql

**Comment by Montellese** [comment]
I'm not sure I understand how this is supposed to work. Up until now I added a source to the music library and didn't have to specify anything (apart from the default artist and album scrapers in the music settings).

Now I have to specify a
* content type? What else apart from "music" is there to choose from?
* scraper? Can I choose both the artist and album scraper? Because AFAIK the music scanner needs both to scan a music source.

File: xbmc/dialogs/GUIDialogMediaSource.cpp [hunk 2]
```diff
@@ -389,11 +390,13 @@ void CGUIDialogMediaSource::OnOK()
   {
     m_confirmed = true;
     Close();
-    if (m_type == "video" && !URIUtils::IsLiveTV(share.strPath) && 
-        !StringUtils::StartsWithNoCase(share.strPath, "rss://") &&
-        !StringUtils::StartsWithNoCase(share.strPath, "upnp://"))
+    if(!StringUtils::StartsWithNoCase(share.strPath, "rss://") &&
```
Comment (Montellese) [inline xbmc/dialogs/GUIDialogMediaSource.cpp]: Space between `if` and `(`

File: xbmc/dialogs/GUIDialogMediaSource.cpp [hunk 3]
```diff
@@ -389,11 +390,13 @@ void CGUIDialogMediaSource::OnOK()
...
+      !StringUtils::StartsWithNoCase(share.strPath, "upnp://"))
```
Comment (Montellese) [inline xbmc/dialogs/GUIDialogMediaSource.cpp]: Indentation

File: xbmc/music/MusicDatabase.cpp [hunk 4]
```diff
@@ -3946,11 +3946,38 @@ void CMusicDatabase::UpdateTables(int version)
...
+      std::vector<std::string> items;
+      // get data from returned rows
+      while (!m_pDS->eof())
+      {
+        items.push_back(m_pDS->get_sql_record()->at(0).get_asString());
+        m_pDS->next();
+      }
+      m_pDS->close(); // cleanup recordset data
+
+      for (auto oldPath = items.begin(); oldPath < items.end(); oldPath++)
```
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]:
```
for (const auto& oldPath : items)
```

File: xbmc/music/MusicDatabase.cpp [hunk 5]
```diff
@@ -3946,11 +3946,38 @@ void CMusicDatabase::UpdateTables(int version)
...
+  {
+      // Convert the content table from using musicdb:// paths to actual paths
+      if (NULL == m_pDB.get()) return;
+      if (NULL == m_pDS.get()) return;
+      //Genres are no longer supported here
+      m_pDS->exec(PrepareSQL("DELETE FROM content WHERE strPath LIKE 'musicdb://genres/%'"));
+
+      m_pDS->query(PrepareSQL("SELECT strPath FROM content"));
+
+      std::vector<std::string> items;
```
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]: Maybe name this `oldPaths`?

File: xbmc/music/MusicDatabase.cpp [hunk 6]
```diff
@@ -3946,11 +3946,38 @@ void CMusicDatabase::UpdateTables(int version)
...
+      // get data from returned rows
+      while (!m_pDS->eof())
+      {
+        items.push_back(m_pDS->get_sql_record()->at(0).get_asString());
```
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]: Why not `m_pDS->fv(0).get_asString()`?

Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: No specific reason, had some problems getting this to work, might have just been the first thing that woked. Will change.

**Comment by evilhamster** [comment]
First of all, I like the idea about setting scanners on music directories and I think it is a nice improvement!

However I think the conversation between old settings and the new one might have some issues. I will try to point them out. 

For albums, if you have an album that have two discs where you save the first disc in album/disc 1 and the second one in album/disc 2. To my knowledge, without a order by statement the returned order of the items is not guaranteed to be the same, so when doing the conversation you might have gotten the disc 1 path first and stored that one and later when you try to rescrape that album you get disc 2 from the sql query. 

For artists, you have the same order problem. If you have the following directory structure:
va/va-album/
artist1/album/
artist2/album/
if the first returned item from the query is the va entry, you would have changed the scraper not only for that artist but for all artists for which the va entry was first returned. Also, if the user removed the va-album, this would also change the scraper being used for the artists affected above. 

If we allow different scrapers on different directories, do we really need/want to be able to change scrapers on a artist/album level? Asking since I'm not using that functionality myself.

**Comment by DaveTBlake** [comment]
@evilhamster 
>If we allow different scrapers on different directories, do we really need/want to be able to change scrapers on a artist/album level? Asking since I'm not using that functionality myself. 

Yes I think we do. The scrapers can get things wrong, it does some funny things with classical music in particular easily mis-identifying albums etc. Then you may want to use that scraper generally but not on a specific artist or album. Also users may turn to local info or a different scraper for some items for this reason.

Sorry @Razzeee I have not had time to look at what this change does in more detail yet. But with @Montellese I am not sure I want to set the scrapers every time I add a source. Maybe use default but allow user to change on a folder basis?

**Comment by MartijnKaijser** [comment]
@Montellese doing it this way brings it in line with videos which was the goal anyway right? Sure there's only "one" music type now but this could make it expandable. Additionally once music and videos a re handled the same you work towards a single "files" section from where you can set any content regardless of video or music/audio.

Indeed handling the scrapers for music is tricky because we now need two scrapers. A solution would be to make a single scraper for albums/artists. I haven't checked current end result on how it looks yet.

**Comment by MartijnKaijser** [comment]
@DaveTBlake why is it a problem to set a scraper on every source? You do this for videos as well. 
If the scraper misidentifies an artist/album then this needs to be solved and not by donig a lot of manual labour to not set content or change scraper for that albums. That's just plain wrong.

**Comment by Razzeee** [comment]
@evilhamster 
this is supposed to only affect the config. I don't think it would change anything with the item orders or scraping itself (other than providing better configs)
But if you can come up with a valid test case, that fails, please go ahead. There might be a corner case, when handling a filestructure in an album folder (like CD1, CD2) when translating an artist to absolute path. Try selecting a two part album in your album library and set a scraper to artists. Try if it can fetch that config, when you directly open up the config again via context menu.

@Montellese 
I would also prefere one scraper, but small steps.

**Comment by Montellese** [comment]
@MartijnKaijser: well IMO music library doesn't have to be identical to the video library. If not showing a dialog with no real use (at least I still don't see one) makes it work a bit different that's perfectly fine for me.

@Razzeee: I didn't say that we should move towards one scraper and I have no idea how feasible that is. I was asking how you handle the fact in the dialog that the user needs to configure two scrapers?

Maybe you could make a screenshot of how the dialog looks like for a music source?

**Comment by MartijnKaijser** [comment]
@Montellese well sorry to say but it's not fine for me. It bothered me for a really, really long time and i'm really happy and thankful this is finally getting tackled. The difference between setting video and music content is totally awkward. Why should you change default settings for music in system setting while it's not there for video? Setting music should become similar to how it's done for video.

**Comment by DaveTBlake** [comment]
@MartijnKaijser since my main use is music library, video I am happy to work from files/folders so have never tagged videos or scanned or scraped a video lib, I don't really understand I what making music behave like video entails. I know there is the wish to integrate music videos, but given the different nature of music I am not sure music users want the libraries to look the same.

All I am saying is have the ability to set the scraper on every source, but for ease (fewer button to press) just default to a default setting initially.

>If the scraper misidentifies an artist/album then this needs to be solved and not by donig a lot of manual labour to not set content or change scraper for that albums. That's just plain wrong.

How does a user "solve" a misidentification? Initially you may not even notice that the scraper has run, deleted artists and replaced them with those of another album. Or it sticks you in an infinite loop because it can't find the album/artist (there is a PR to fix this longstanding bug).There are users that manually use sql to set the scraped date to stop the scraper all together. When scraping works it is invisible and wonderful, but when it doesn't it is a user nightmare that many can't even work around. That I plain wrong I agree, but untill we can be sure the scraper works 100% we have to leave users a way out.

**Comment by MartijnKaijser** [comment]
@DaveTBlake i'm talking about making the code better to prevent misidentification. If those users come to the forum and start a discussion on where the scraper fails and there's no music library dev who can look into it will remain as is. However that is all off topic and doesn't have anything to do with actually setting content on a music folder which this PR is about.

**Comment by Montellese** [comment]
@MartijnKaijser: just showing a dialog for the sake of consistency even when the dialog doesn't offer anything useful just doesn't make any sense to me. If you change the dialog to allow choosing both scrapers and their settings that's fine by me but only being able to configure half of the settings (i.e. only one of two scrapers) just gives the user the wrong impression that he has fully configured the source even though he hasn't.

Maybe I'm misunderstanding something here which is why I asked for a screenshot.
Right now to me it looks like the user has to configure every artist and every album separately (which would be a loooooot of work) because the dialog only allows to choose between album and artist even though every normal music source will contain both artists and albums. That's what makes it so different from video sources where we have one scraper for movies, one scraper for tvshows, seasons and episodes and one scraper for musicvideos. Abusing the same dialog to configure a music source which requires setting two scrapers just doesn't feel right.

**Comment by MartijnKaijser** [comment]
@Montellese sorry but i will press this change forward. I am one of the few that even uses music so i'd say trust that judgement. All i will ask is that others review the code changes.

**Comment by Montellese** [comment]
@MartijnKaijser wow that's a very good argument. I'm out of this discussion. But don't ever come back asking for help on this from me.

**Comment by MartijnKaijser** [comment]
@Montellese why such a response? I am not trying to offend you but i use music daily and am only trying to make the flow of getting it configured more consistent and easier like video. As @Razzeee already said we need small steps at a time.

**Comment by Razzeee** [comment]
image

image

Merging artist and album scrapers to one music scraper would be the next logical thing. But I think thats over my head, as long as we're still using the this type of scrapers.

This might also pave a way of a global file/dir listing. Like the video item now, minus it only working for videos, we could use it for all configuration that is needed at staring with your librarys. One central point. Select a folder, now I just have to choose if this is music, a tv show or a movie.

**Comment by Montellese** [comment]
@MartijnKaijser: Because you're not providing any arguments (apart from "music and video have to work the same"). I have explained my PoV on why I think this is wrong and how it should be improved but you said you'll ignore it so I don't see why I spent the time to write my PoV and think about how to improve it.

**Comment by jjd-uk** [comment]
I need to try these changes out but on balance I think I agree with Martijn, we should try and aim for consistency in how sources are added & managed, as long as it doesn't make for a poorer user experience in the Music section.

My thinking is it might help expansion to having an audio library rather than just a music library, perhaps we might want to add other stuff such audio book support which would need a different scraper.

**Comment by MartijnKaijser** [comment]
@Montellese sorry i thought i mentioned the arguments already. Maybe in a not so clear way appearently.

Here's a scenario:
I have one big music folder for which i want a different than default scraper. Just added the source, need to go to system->music->library->pick different default scrapers. Go back to source, scan to library. Wait till it's done. Go back to system->music->library->pick different default scrapers, go back to one of the subfolders, scan to library to use the different scrapers.
Find out that one of the albums went wrong in that subfolder and need different scraper, go back to library settings to choose different scraper. 
Not to mention that if i want to change the setting of that scraper i need to go to add-on manager, change the scraper setting and go back to the source window.
Doesn't really sound that user friendly to me.

Proposed way as done here:
Add source, set content, pick album scraper (perhaps change the settings), pick artist scraper (perhaps change it's settings)-> scan library. All this from within same place in Kodi.
If you find out that you need different scraper on a sub folder you go there and repeat above process just like for videos.

Yes setting the album and artist scraper may not seem that intuitive but it's a lot better than how it's now.

**Comment by Montellese** [comment]
@MartijnKaijser: The fact that you should configure the album and the artist scraper in the same dialog is mentioned nowhere. I was thinking under the assumption that the dialog lets you either choose "albums" OR "artists" as a media type and a matching scraper. That would make no sense which is why I explained above because a normal music source will contain artists and albums.

The problem I (and you apparently as well) see with the explained approach is that from the video library people are used to choose one type and configure one scraper and now we try to make it consistent but it works different in an essential way i.e. you have to know that you don't choose a media type but instead you switch between media types to configure them. Yes, it's a lot easier than going back and forth between the music files view and the music settings but it's also very unintuitive.

As seen in the screenshots above we have a lot of unused space at the bottom. Couldn't we somehow show both scraper selections at the same time i.e. the artist scraper at the top and the album scraper at the bottom?

**Comment by MartijnKaijser** [comment]
@Montellese ok sorry then i should have explained it better (sorry for pushing my opinion). Lets blame wrong assumptions.
Perhaps not using "artist/album" as content type but just music and giving the user the dialog twice after each other which gives the option to define scraper and making it clear to the user that it's for the same source.

@Montellese one of the possibilities i mentioned to @Razzeee yesterday was to perhaps set "music" as content and use the space below for selecting the default scraper. So move the "default scraper" setting from system->music to the part below the selection where videos has the options. Though this makes it kinda weird.
(perhaps not explained in a that clear way).

So yes music is different from videos but that's just because it has two scrapers while movies/series only has one. Merging the two album/artist into one scraper would be the most ideal situation that solves this whole pickle.

Also the empty space below might look weird here but switching skins might give a totally different appearance.

**Comment by jjd-uk** [comment]
Can we have tabs in the Set Content dialog? so having an Artists tab and an Album tab might make it clearer that you set scrapers for both.

**Comment by Razzeee** [comment]
@MartijnKaijser 
in fact switching scrapers already shows something below, if I remember correctly. So it's up to the scraper.

I honestly think the only way to go is merge artist and album scraper. Maybe @olympia can give us a clue how hard that will be.
The thing I'm fearing is, we might depreciate third party scrapers with this move.

On another note, refactoring music will lead us needing to make more hard decisions. And I think we will have more problems where we need to depreciate something to be able to move along.

**Comment by jjd-uk** [comment]
Would it be possible if someone was so inclined to add an "Exclude path from library updates" under "Content scanning options" similar to what there is for videos? if so then that might be another argument for using the "Set content" dialog as this would allow subfolders to not be added to the library, for example you might have a subfolder under your main Music folder which stores poorly tagged music.

**Comment by evilhamster** [comment]
@Razzeee 
Well one use case, empty isengard installation where you add one VA album and change the scraper for two artists. The content will the look like:
musicdb://artists/2/|metadata.local|artists|
musicdb://artists/13/|metadata.local|artists|
Update to this version and it looks like:
/home/snapper/test/music/VA/|metadata.local|artists|
/home/snapper/test/music/VA/|metadata.local|artists|

This would have the effect that all artists (since I only have the VA album added) would use the local scraper, but there seems to be something wrong and when I try to refresh any artist it is using the fallback.

In regards to ordering, that is harder to show since ordering is based on the query optimizer and storage implementation, if they are the same the order should be the same but if they change the order returned without an order by could change.

**Comment by DaveTBlake** [comment]
Can I just clarify how this changes the user experience (allowing for the fact that I am not familiar with video, and my attempt at it 5 mins ago was a bust - so video not so intuitive either!).
I am onboard with moving the scraper selection nearer the file view thus making switching scraper settings easier. But I am having a language problem.

At present scanning the tag info from music files into the library is a separate thing from scraping additional information about artists or albums. Scanning is on the context menu from the file list (having added it as a music source), for on library update.  Scraping is initiated either from the context menu of an album or artist item  by either "query info for all..." or looking at the information for that artist/album. In the switch to a "set content" approach do we loose this separation? It seems to mix scanning options with scraping ones?

I have a broad music collection. Scraping is OK for pop music, but the results are not so reliable for classical music or more obscure titles. I Like the idea of not scraping some folders, or not scraping at all,  but still want to scan the tag information. But I can see it could be the other way around if you have not tagged well. Scanning and scraping are different things right?

**Comment by Razzeee** [comment]
@jjd-uk 
adding this to the scraper page /database won't be a problem, I have to check the integration into the scraper itself. Haven't touched that so far.

@evilhamster 
I'm still not getting the theory your hunting with that order topic.
And yes, that is a corner case, not sure if we can do anything about something like that.

**Comment by jjd-uk** [comment]
@Razzeee yeah sure, just wondered if it was something that might be possible now with using this approach to adding music content. So that would be another argument for using a dialog like this which might help @Montellese understand why it could be an improvement and what future options it might give us.

**Comment by zag2me** [comment]
Just wanted to say as a regular music user this has bugged me for a while also. Its nice to see changes like this making the interface more consistent. It basically makes it easier to import your music which is an improvement.

**Comment by jjd-uk** [comment]
Btw I've just kicked off a jenkins build to upload test builds to the mirrors, will update here when done.

**Comment by evilhamster** [comment]
@Razzeee 
Because you are using the first entry of the select statement as a key to identify the scraper for that artist / album. If the select statement returns more than one entry, the order matters. Example, let say I have an artist that exists on two albums, that are located on different places on my share:
_VA/Album1/
Artist/Album2/

Let say the content table looks like:
 /VA/Album1|metadata.local|artists|

 If I later on upgrade my db or do something else that effect the order that the sql return the items in, it will also change which scraper is being use for the artist in question (because then it will try and find Artist/ in the content table). 

An other problem is that if I decided to remove the VA album, I will also change the scraper that is being used for the artist. 

Also if I have a structure like:
va/album1
va/album2
artist1/album
artist2/album4

My only point is in using one files parent location as a key for an artist can cause problems down the line.

**Comment by evilhamster** [comment]
I think the problem could be solved if we handled the scrapers configuration like the following.

We have a list of scrapers that should used for different locations, like in this pr in content. For artists/albums where the users have manually selected to use a different scanner, we save that information as well but that information we save as reference to the artistid. So the logic when rescanning an artist would be something like:
a) Check if there is an override for this artist, if it does scan using custom scraper
b) If not, use the default scraper for the most common location where that artist exists.

**Comment by Tolriq** [comment]
From an heavy music user too point of view, the current behavior of 2 scrapers in settings and not on source is complex and not intuitive at all.

I'm quite used to Kodi since all those years, and yet I often made mistakes when building a new library for tests or friends.

That said : The current behavior for English speaking people that listen to basic commercial music, is very easy to new users.

Having them to handle 2 different scrapers for a single source would lost most of them. So not sure 2 dialogs is the way to go.

IMO a much more logical way would be at source add : A dialog asking use default online scrapers or local information only.
Then on right click on the source : Set album scraper + Set artist scraper.

(Of course a merge of the scrapers would be cool, but I'm afraid it would not covers all needs, for foreigners that user non standard artist scrapers but still scrape albums with default one).

**Comment by phil65** [comment]
My WIP stuff already contains a much more flexible layout for DialogContentSettings.xml which moves scraper selection to DialogSelect. That would be a first easy step to get things into more basic and more flexible layouts and would, if used for music somewhen, allow an additional button to have a second scraper selection button (artists+albums)
Merging album and artist scraper doesnt make sense to me at all.

**Comment by MartijnKaijser** [comment]
What rush? There is no rush? This is still WIP and solving it in a skin is not the solution. Merging album and artist scraper makes perfect sense. Data nearly always comes from the same website.

**Comment by phil65** [comment]
> sorry but i will press this change forward.

That rush.

This isnt shot down, but there are some valid concerns brought up by @Montellese and @DaveTBlake (scraping vs. scanning for example) which seem to just get ignored. I dont want to end up with stuff which is not properly thought trough.

The same website as a source is not a valid reason to merge scrapers. Or would you also merge Movies and TVShow scraper if they are delivered from the same website? (like tmdb for example)

**Comment by MartijnKaijser** [comment]
ok then i now request separate scrapers for series, seasons and episodes

**Comment by DaveTBlake** [comment]
@MartijnKaijser 
>What rush? There is no rush? 

Glad to hear it, a measured approach will have a better outcome. The tone of your earlier posting  where you said 

> i will press this change forward. I am one of the few that even uses music so i'd say trust that judgement. All i will ask is that others review the code changes

it seemed that it was all happening today, and all you wanted was others to check for typos. Well at least it has sparked discussion even if you don't want it.

**Comment by MartijnKaijser** [comment]
To clarify the "scraping vs. scanning" concerns are void as that is not touched. All this attempts to do is replace the "scan to library" context menu and set a default scraper to that source. You still have manually trigger scan for more information through context menu yourself for artists and albums

**Comment by phil65** [comment]
It is not void. If you define the scraper(s) per source then the global "additional-info" setting would be obsolete I think? Since it would allow then to choose to not use any scraper at all for a source when adding it to the library.

**Comment by Razzeee** [comment]
@phil65 
we could rip the default out.
But I was thinking more along the lines of getting this rolled out. If they set up a new system or just new librarys, they'll migrated to the new way. Then depreciate the old way in some future version. Nice and easy.

**Comment by MartijnKaijser** [comment]
The mentioned concerns where that it would start automatically pulling in unwanted info. That had not changed. What you mention now is another point on moving the scan for extra info to the set content window

**Comment by DaveTBlake** [comment]
@Razzeee 
Can you do anything to avoid the user confusion between scanning and scraping?

At the moment we add a source, the use context menu to scan (the tag data) into the library. Subsequently we may deliberately scrap additional artist/album info for all artists/albums or inadvertently for a single album/artist by attempting to look at the info.

The proposal here is that we add a source and at that moment are asked to choose scrapers and if we want to scan. The implication is that the scrapers are used in scanning, but I believe that scraping is, and should continue to be, a separate activity to scanning.

**Comment by Razzeee** [comment]
Not sure how I could avoid that confusion. I'm also confused and also a little bit shocked, by the fact, that we're not using the scrapers when scanning. Why would we try to leave the user with just tags, when he can have an enriched media experience.
Thats probably just another design flaw of an historically grown system. Which we should fix asap, if you ask me. 

Normal users should not care, if Kodi is scanning or scrapping. All they care about is, good data and probably a beautifuly display of whatever is running.

**Comment by MartijnKaijser** [comment]
@DaveTBlake indeed that scan vs. scrape is very confusing. If that can be moved to the bottom section per source for "scan for additional information online" that would at least make it more clear. Maybe some extra line in there that says scanning will only use tag data.
So you could change it for your pop and classical music separate.

**Comment by Tolriq** [comment]
For information if that can help, I had a lot of request from users to implement tons of things for music in file mode.

I asked them all why dont you use the library as it opens tons of things and each time the answer was Kodi does not scrape correctly my data so I prefer using my folder organisation.

Often root/genre/artist/albums.

And from the stats I have for Yatse users there's 3 more times files played from file mode than library for music.

So maybe this scan vs scrape thing is important to better explain if scrape can override tags things like genres.

**Comment by phil65** [comment]
Why should we add a separate button to the bottom section for "scan for additional information" if the same dialog would contain scraper selection? That button basically is an on/off toggle for scraping. 
If a scraper is selected it should automatically scrape imo, if people want their niche music to not get scanned then they would just choose "no scraper".

**Comment by Razzeee** [comment]
@phil65

**Comment by DaveTBlake** [comment]
@phil65 
>if people want their niche music to not get scanned then they would just choose "no scraper".

That is just the kind of confusion I want to avoid. I very much want my classical music __scanned__, I want to use the quality tagging I have applied, but I don't want to let the scraper mess it up so I would choose "no scraper", or whatever to stop __scraping__.

What Tolriq says is very significant. The results of the scraper (not scanning) can be unwanted, and they can happen unbidden, consequnetly users give up on library entirely!!

@Razzeee 
I am sure some users would agree with you, it all depends on what music you listen to and how you listen to it. But there is also a number of us, quite a number from Yatse stats, that don't have the TV on when listening so really don't care about pretty fan art etc. or biog info. We tag our music well, we organise it on our drives well and just want Kodi to make selecting what to play really easy yet flexible.

Please don't make scraping compulsory!!

But you are right, most users don't care about scaning or scraping (until it goes wrong), and shouldn't need to.

**Comment by DaveTBlake** [comment]
@Razzeee 
> I'm also confused and also a little bit shocked, by the fact, that we're not using the scrapers when scanning. 

I have to say I am a little worried that you have got this far with this PR and not realised that scraping (extra info from online data sources) is different to scanning (tag data from the music files). 

There will always be some of us that have more faith in our local data over the "wisdom of crowds" out on the web. Some things like genre are subjective and down to personal choice anyway. Scraping can be powerful and effective, and made my pop music library look very pretty when I accidently clicked on "query info for all.." for the first time,  but it is not best for every situation.

>Why would we try to leave the user with just tags, when he can have an enriched media experience.
Thats probably just another design flaw of an historically grown system. Which we should fix asap, if you ask me. 

I don't see user choice as a flaw. Some of us are very anal about our music collections, we don't want external input messing it up. One person's "enrichment" is another's nightmare.

What an interesting PR this has become

**Comment by phil65** [comment]
@DaveTBlake : sorry, my wording was wrong :) I meant "scraped", not "scanned".

**Comment by jjd-uk** [comment]
The jenkins build all failed so no test builds uploaded so I've not had chance to check this out yet.

I notice in the screenshot there's a local information only scraper, is there anything in place behind this? I ask because if someone only want to scan their music files tags and not scrape online info then this would seem to be the logical choice.

However if they've selected a scraper then as phil65 has stated, this implies online metadata is wanted.

**Comment by DaveTBlake** [comment]
Yes, I'll agree with @phil65 and @jjd-uk selection of a scraper (for a source), with an option of none, could be an on/off toggle for scraping. It would also be good to warn the user, as @Tolriq suggests, what scraping does e.g. overwrites some tag data,  compared to scanning that constructs the fundamentals of the library.

Even better would be the user facility  to audit what changes and additions the scraper wants to make and then accept or reject them. As I said some of us are control freaks over our music library!

**Comment by Razzeee** [comment]
I'm really really afraid most of you are talking without taking everything into account and doing a little research.
As I stated above, this just unifies a feature that was already there. I didn't change anything on the scraper part. Even if we would scan and scrap by default, we would not overwrite any tags on the fies (as long as you don't activate that setting in the music settings). I think most users are not even aware that we're able to additnally scrap. I always thought Kodi was taking this out of my hands, but today I learned I was wrong.
If you as audiophile don't want it fine, disable it. But don't force your niche use case onto all users

**Comment by Montellese** [comment]
Maybe the content dialog should only be shown if the "download additional information" setting is enabled?

**Comment by Razzeee** [comment]
At this point I'm like "why do we even have scrapers if everybody just seems to want them to go away?"
Scanning doesn't deserve an own name, its just scraping from local files :P

**Comment by MartijnKaijser** [comment]
Maybe the content dialog should only be shown if the "download additional information" setting is enabled?


^^ then you still end up going back and forward between settings and files because you do or for not want extra info scraped.

**Comment by phil65** [comment]
@Razzeee it would be nice what you think was so wrong with those comments before turning into desparation.
The situation for videos cannot get transferred 1:1	 to music because music can also populate the library without getting scraped.	That is quite a big difference which atm is reflected by an additional setting in backend and a different way to add stuff to the library. (I think that´s also what Montellese tried to point out earlier)
That said, it doesnt seem impossible to me and some others here that the workflow can be unified to a great extent by also makin use of the content dialog.
The proposal which came up then was to show the content dialog for music folders including two scraper choice settings. If scraper(s) is set to "None", then it should just take the metadata from the file itself. If a scraper is chosen, it should scan metadata from file + scrape additional info from remote sources (which, as @DaveTBlake correctly pointed out, overwrites the database entries and therefore the infolabel output, and of course not the metadata from the file itself)	
That way the backend setting could get removed, people can add their music folders to library with scrapers on a per-folder base just like for videos and all people are happy.

**Comment by Razzeee** [comment]
No body would want a 1:1 experience. You just shoul be able to work out how the one works, if you know the other. And the other way around.
It's different but not so different. If we think something is worthwhile in video, it's likely that something like that may also have value in music and the other way around.

Here is my evaluation, to be fair I like both.

- Option one like you said two options to select artist/album scraper:
   - Breaks away from standard that video uses as it has two content types. Might confuse the user.
   - Setting one of the items to None would disable the other one. So we also need to gray it out and get some code into place, that disables switching then etc. While this is possible, I think this may just end up in a very clunky and hard to understand user experience.

- Option 2 replace artist/album scraper by music scraper
    - Merge matching artist and album scrapers into one music scraper (so universal album and artist = universal music) Add some nice settings for the scraper and every video user will know what he needs to do.
   - Less items to edit for the average joe user, which just wants to setup scraping for his database
 
Also please think about the future and the possibility to have one files listing for all folders, where you can set content types. It's quiet an overlooked problem right now, you need to absolutely know what kind of data you want to edit. You can't see all your scraped folders (video, music, musicvideos, pictures(?)) at once. If you want to edit something, you need to know where to go, experienced users will know that. Rookies won't.

**Comment by phil65** [comment]
ok, so we´re finally back on track.
I admit that I was wrong when saying that a generalized music scraper doesnt make sense at all. Both outlines you suggested sound like something which would be an improvement, at least to me. I still somehow feel that we limit ourself in an uneeded way somehow by merging scrapers, but if people dont want that grade of granularity, then fine.
I would still consider some input and a confident "Go for it" from one of the "senior devs" as really essential here. Questioning the knowledge of all people involved really doesnt help for that. ;)

**Comment by DaveTBlake** [comment]
Can we talk some more about the user experience (that we are trying to improve).

At present when I add a music source I have a context menu item to "Scan to library", and it is obvious to me first time that I want to do that to create a music library. Having done it once doing it for additional sources makes sense too.

When I add a video source  the "Set Content" menu item means nothing to me, so I have never created a video library, I just use the file view. 

We all seem to agree that it is an improvement to move the music scraper settings nearer to the file view. But it is far from clear to me that replacing music menu "Scan to library" with  "set content" is better for users. Perhaps video could benefit from mentioning "library" in the menu item?

The workflow is also significantly different. In the video case to populate the library we set scrapers for the kind of material it is and use them to fetch external data. You need to be online to produce a library. In the music case to populate the library we scan the tag data from the music files themselves. This can be done without an internet connection. We can also augment/amend the data with artist/album info scraped from external sources, but it is not essential. These differences are inherent in the material.

Whatever UI we end up with we need to be able to populate the music library with tag data (off line), and then go and scrap extra data later. Default to scanning tag data and scraping external info all at once if you want, but allow for them to be separate actions.

The workflow is different.

**Comment by Tolriq** [comment]
Just to repeat myself in case it was not so visible in all those messages :

Unified scraper should really be thought globally and not just for English speakers.

I for myself scrape the albums with Kodi to have additional information on them since my tags are not perfect, and I do not really bother that the information are in English.

But for artist I like to have great controlled data in French.

And from current scrapers if merged this won't be possible and quite a big problem forcing me to generate NFOs for my albums (3000 ... without any real good tools like Ember to achieve that) this would be a total mess.

And I'm sure I'm not the only one in that case.

But this is to me a more global issue about scrapers and NFOs, currently AFAIK you can either use NFO or scraper or both but with scraper overriding NFOs.

Maybe an option to have NFOs > Scraper would solve many problems too for all the data centric maniacs that can't manage everything so need scrapers but needs controls on some part without the need to create tons of sources to separate things.

**Comment by DaveTBlake** [comment]
Oh yes, in a variant of @Tolrig I would use the artist scraper but not the album one. It can correctly identify Beethoven, but has trouble differentiating different albums of Symphony No. 5 etc. So users do need to be able to scrap one kind of metadata and not the other.

NFO overrideing scraper is also an excellent idea for all the reasons he mentions.

**Comment by Razzeee** [comment]
>The workflow is also significantly different. In the video case to populate the library we set scrapers for the kind of material it is and use them to fetch external data. You need to be online to produce a library. In the music case to populate the library we scan the tag data from the music files themselves. >This can be done without an internet connection. We can also augment/amend the data with artist/album info scraped from external sources, but it is not essential. These differences are inherent in the material.

Videos is the absolutely the same, you can scrap offline from nfo. Which is essentially just what music librarys scanning does.

**Comment by Tolriq** [comment]
@Razzeee as explained this is wrong : There's TAGS, NFO and scrapers.

Music with current handling have limitation in what the user can choose do with those 3 cases.

NFO is scraping not scanning. TAG is scanning. Scrapers are scraping.

A simple way to content every one I suppose would really to have some better ordering of the priorities.

Can NFO overwrite TAGS ? (This one is rhetorical since logically NFOs are here just for this purpose but maybe there's user case here)
Can Scrapers overwrite TAGS and / or  NFOs ? (This is the big question that if solved correctly can satisfy all the needs)

If you can configure things to have TAGs and or NFO not overridden by scraping then merged scrapers are a possibility since user have a way to control the data he want, and the one he does not want.

**Comment by notspiff** [comment]
no, nfo cannot override tag. they are complementary. nfos are not there to supply the information supplied in tags, but to supply information that isn't there - rich info about artist and albums. it is nfo and scrapers which provide the same information.

**Comment by DaveTBlake** [comment]
>Videos is the absolutely the same, you can scrap offline from nfo. Which is essentially just what music librarys scanning does.

No that is totally incorrect. Scanning populates the library with fundamental music data from tags about artist, album and song. Scraping and NFO files (only useable if you have scraped) provide additional artist and album info, nothing about the songs (the song list for an album is not connected to the songs you actually have in the library).

**Comment by Razzeee** [comment]
Discussion moved here: http://forum.kodi.tv/showthread.php?tid=239380&pid=2110044#pid2110044

**Comment by Razzeee** [comment]
jenkins build this please
@jjd-uk mentioned this failing on other platforms, I would like to know why

**Comment by Razzeee** [comment]
Let's see if their okay now
jenkins build this please

**Comment by hudokkow** [comment]
Needs rebase.

**Comment by MartijnKaijser** [comment]
@hudokkow why? It's not going in as is

**Comment by Razzeee** [comment]
Rebased and squashed as I was bored.
